Source/AutoExtern/Program.cs

```diff
@@ -0,0 +1,414 @@
+  }
+
+  public Name IntfName => Name.OfSyntax(syntax, model);
+
+  private bool IsInterface => syntax.Keyword.ToString() == "interface";
+
+  private IEnumerable<PrettyPrintable> Fields =>
+    syntax.ChildNodes().OfType<FieldDeclarationSyntax>() // Filtering for `public` fields is done later
+      .Select(s => new Field(s, model));
+
+  private IEnumerable<PrettyPrintable> Properties =>
+    syntax.ChildNodes().OfType<PropertyDeclarationSyntax>().Where(v => model.IsPublic(v))
+      .Select(s => new Property(s, model, IsInterface ? this : null));
+
+  public override void Pp(TextWriter wr, string indent) {
+    PpBlockOpen(wr, indent, "trait", IntfName,
+      syntax.TypeParameterList?.Parameters.Select(s => new Name(s.Identifier).DafnyId),
+      new Dictionary<string, string?> { { "compile", "false" } },
+      syntax.BaseList?.Types.Select(t => new Type(t.Type, model)));
+    PpChildren(wr, indent, Fields);
+    PpChildren(wr, indent, Properties);
+    PpBlockClose(wr, indent);
+  }
+}
+
+class Enum : PrettyPrintable {
+  private readonly EnumDeclarationSyntax syntax;
+  private readonly SemanticModel model;
+
+  protected override string ChildSeparator => "";
+
+  public Enum(EnumDeclarationSyntax syntax, SemanticModel model) {
+    this.syntax = syntax;
+    this.model = model;
+  }
+
+  private IEnumerable<EnumMember> Members =>
+    syntax.Members.Select(m => new EnumMember(m, model));
+
+  public override void Pp(TextWriter wr, string indent) {
+    var nm = Name.OfSyntax(syntax, model);
+    PpBlockOpen(wr, indent, "class", nm, null, null, null);
+    PpChildren(wr, indent, Members);
+    PpChild(wr, indent, $"function method {{:extern}} Equals(other: {nm.DafnyId}): bool");
+    PpBlockClose(wr, indent);
+  }
+}
+
+
+class EnumMember : PrettyPrintable {
+  private readonly EnumMemberDeclarationSyntax syntax;
+  private readonly SemanticModel model;
+
+  public EnumMember(EnumMemberDeclarationSyntax syntax, SemanticModel model) {
+    this.syntax = syntax;
+    this.model = model;
+  }
+
+  public override void Pp(TextWriter wr, string indent) {
+    var decl = Name.OfSyntax(syntax, model).AsDecl(forceExtern: true);
+    var type = Name.OfSyntax(this.syntax.Parent!, model).DafnyId;
+    wr.WriteLine($"{indent}static const {decl}: {type}");
+  }
+}
+
+class Field : PrettyPrintable {
+  private readonly FieldDeclarationSyntax syntax;
+  private readonly SemanticModel model;
+
+  protected override string ChildSeparator => "";
+  protected override string ChildIndent => "";
+
+  public Field(FieldDeclarationSyntax syntax, SemanticModel model) {
+    this.syntax = syntax;
+    this.model = model;
+  }
+
+  private IEnumerable<Variable> Variables =>
+    syntax.Declaration.Variables.Where(v => model.IsPublic(v))
+      .Select(v => new Variable(syntax.Declaration.Type, v, model));
+
+  public override void Pp(TextWriter wr, string indent) {
+    PpChildren(wr, indent, Variables);
+  }
+}
+
+class Property : PrettyPrintable {
```

Comment:
That's a good point.  I have the opposite issue: my editor isn't fancy enough, so if everything is in one file I can trivially search around, rename functions, search and replace, etc; whereas otherwise I need advanced support for navigation and refactoring.  Same for errors: if I change something I can see everything that breaks from the change in one file, rather than having to visit them all or recompile.

(Also, I usually dislike splitting because it loses history, but since our merge process already does that it's not an issue here.)

Still, I wouldn't mind splitting this one: would you just move the abstract class and keep all the nodes together, or would you do one file per node class?

Reply:
Given that they're large enough to fill a screen, I'd put the abstract class and subclasses in separate files. However, I recognise this is subjective. I can only mention this was the norm in the two C# dev teams I've worked in, both of which were using Visual Studio with ReSharper as an IDE.